realm/realm-library/src/main/cpp/io_realm_internal_OsObjectStore.cpp

```diff
@@ -0,0 +1,142 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "io_realm_internal_OsObjectStore.h"
+
+#include <object_store.hpp>
+#include <shared_realm.hpp>
+
+#include "util.hpp"
+#include "jni_util/java_exception_thrower.hpp"
+#include "jni_util/java_exception_thrower.hpp"
+
+using namespace realm;
+using namespace realm::jni_util;
+using namespace realm::util;
+using namespace realm::_impl;
+
+// FIXME: Enable after https://github.com/realm/realm-object-store/pull/550 merged
+//static_assert(io_realm_internal_OsObjectStore_SCHEMA_NOT_VERSIONED == static_cast<jlong>(ObjectStore::NotVersioned),
+//              "");
+
+JNIEXPORT void JNICALL Java_io_realm_internal_OsObjectStore_nativeSetPrimaryKeyForObject(JNIEnv* env, jclass,
+                                                                                          jlong shared_realm_ptr,
+                                                                                          jstring j_class_name,
+                                                                                          jstring j_pk_field_name)
+{
+    TR_ENTER_PTR(shared_realm_ptr)
+    try {
+        auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+        JStringAccessor class_name_accessor(env, j_class_name);
+        JStringAccessor pk_field_name_accessor(env, j_pk_field_name);
+
+        auto table = ObjectStore::table_for_object_type(shared_realm->read_group(), class_name_accessor);
+        if (!table) {
+            THROW_JAVA_EXCEPTION(env, JavaExceptionDef::IllegalArgument,
+                                 format("Class '%1' doesn't exist.", StringData(class_name_accessor)));
+        }
+
+        if (j_pk_field_name) {
+            // Not removal, check the column.
+            auto pk_column_ndx = table->get_column_index(pk_field_name_accessor);
+            if (pk_column_ndx == realm::npos) {
+                THROW_JAVA_EXCEPTION(env, JavaExceptionDef::IllegalArgument,
+                                     format("Field '%1' doesn't exist in Class '%2'.",
+                                            StringData(pk_field_name_accessor), StringData(class_name_accessor)));
+            }
+
+            // Check valid column type
+            auto field_type = table->get_column_type(pk_column_ndx);
+            if (field_type != type_Int && field_type != type_String) {
+                THROW_JAVA_EXCEPTION(
+                    env, JavaExceptionDef::IllegalArgument,
+                    format("Field '%1' is not a valid primary key type.", StringData(pk_field_name_accessor)));
+            }
+
+            // Check duplicated values. The pk field must have been indexed before set as a PK.
+            if (table->get_distinct_view(pk_column_ndx).size() != table->size()) {
```

Comment:
Should Object Store's `validate_primary_column_uniqueness()` be exposed to bindings?

Reply:
I guess it is fine, it is only one line of code + we already have the column index :P